### PR TITLE
Fixed #28385 -- Avoided ignoring natural keys when primary key has a default value.

### PR DESCRIPTION
**Pull request by dmytrostriletskyi**
## Ticket

https://code.djangoproject.com/ticket/28385

## Bug space

When a models primary key has default set, and a fixture is loaded that does not have a pk attribute, then the serialisation code will refuse to update the existing instance and instead try to create a new record in the db.

When the primary key attribute has a default value - such as models using UUID primary keys - the model will fill the pk attribute with that at instantiation time, and the check in build_instance will see that default value and refuse to goto the database to fetch the existing instance.

This is against the expectations set in the docs, that when the model/manager implement the natural keys interfaces and the primary key is omitted, the framework will update the existing instance when deserialising.

(c) reporter Daniel Knell

## How to reproduce

1. Create the same model like [there](https://github.com/django/django/pull/8801/files#diff-2cf99154fd579c9e2d2ee0945f90ef4bR26).

2. Make `loaddata` fixtures flow like [there](https://github.com/django/django/pull/8801/files#diff-79e3f82c439838acf1377e2ec7e1342bR79).

We expect, that if `loaddata` dictionary has no `pk`, it will get object by `natural key` and update object, but it raises:

```
django.db.utils.IntegrityError: Problem installing fixture 'filename.json': 
Could not load app.Model(pk=<pk_number>): UNIQUE constraint failed: app_model.some_field
```
## Why it happens 

1. [Deserializer](https://github.com/django/django/blob/master/django/core/serializers/python.py#L98) flow does not provide `pk` for objects without it.
2. `Deserializer` try to get `pk` from `natural_key` in [build_instance](https://github.com/django/django/blob/master/django/core/serializers/base.py#L215) method.
3. But this method create new model object with default `pk` before implement point `2` [there](https://github.com/django/django/blob/master/django/core/serializers/base.py#L222).

Again: when model is created, we can get `pk` of this model (because `pk` has default value).
Question: how we can get model's `pk` by `natural key` from model without `pk`, if we create one instance before with default `pk`?

## Solution

I got `build_instance` logic and move it to previous level within [Deserializer](https://github.com/DmytryiStriletskyi/django/blob/8d2564a2805f2ebe48a9e5194b5c678f016f4e59/django/core/serializers/python.py#L79) to suitable place.

If it seems okay, we can go ahead and rewrite logic for [that](https://github.com/django/django/blob/d0f59054d0c4b2291a4e0e94ec00537f98a4ab50/django/core/serializers/xml_serializer.py#L213). Also, we can change `build_instance` logic for this two cases instead of touch serializers flow. 

It was not possible to use `TDD`. May add tests in last commit.
Thank you for paying attention.

### CODE REVIEW

File: django/core/serializers/python.py [hunk 1]
```diff
@@ -100,6 +100,21 @@ def Deserializer(object_list, *, using=DEFAULT_DB_ALIAS, ignorenonexistent=False
                 data[Model._meta.pk.attname] = Model._meta.pk.to_python(d.get('pk'))
             except Exception as e:
                 raise base.DeserializationError.WithData(e, d['model'], d.get('pk'), None)
+        else:
+            # If the model instance doesn't have a primary key and the model supports
+            # natural keys, try to retrieve it from the database.
+            default_manager = Model._meta.default_manager
+
+            if hasattr(default_manager, 'get_by_natural_key') and hasattr(Model, 'natural_key'):
+                natural_key = Model(**d['fields']).natural_key()[0]
```
Comment (orf) [inline django/core/serializers/python.py]: Natural keys can be tuples of any length, selecting just the first element isn't always valid I think?

Comment (DmytryiStriletskyi) [inline django/core/serializers/python.py]: > Natural keys can be tuples of any length, selecting just the first element isn't always valid I think?

Yes, you are right! I fixed it.

**Comment by DmytryiStriletskyi** [comment]
Can I follow `DRY` rule and use `ddt` (does not exist in `requirements`) for tests?

File: tests/serializers/test_natural.py [hunk 2]
```diff
@@ -67,6 +67,36 @@ def natural_key_test(format, self):
     self.assertIsNone(books[1].object.pk)
 
 
+class NaturalKeyWhenPkHasDefaultValueTestCase(TestCase):
+    """
+    Verify that deserializer does not ignore natural keys when primary key has a default value (ticket `28385`).
```
Comment (MarkusH) [inline tests/serializers/test_natural.py]: Please wrap docstrings at 79 chars.

File: tests/serializers/test_natural.py [hunk 3]
```diff
@@ -67,6 +67,36 @@ def natural_key_test(format, self):
...
+        """
+        Test fixture based on `json` format.
+        """
+        management.call_command("loaddata", "fixture.json", verbosity=0)
+        self.assertEqual(Foo.objects.all().count(), 1)
+
+        management.call_command("loaddata", "fixture.json", verbosity=0)
+        self.assertEqual(Foo.objects.all().count(), 1)
+
+    def test_natural_key_xml(self):
```
Comment (MarkusH) [inline tests/serializers/test_natural.py]: Is there any reason why you test the same for both JSON and XML?

File: tests/serializers/test_natural.py [hunk 4]
```diff
@@ -67,6 +67,36 @@ def natural_key_test(format, self):
...
+    """
+
+    fixtures = ['fixture.json', 'fixture.xml']
+
+    def test_natural_key_json(self):
+        """
+        Test fixture based on `json` format.
+        """
+        management.call_command("loaddata", "fixture.json", verbosity=0)
+        self.assertEqual(Foo.objects.all().count(), 1)
```
Comment (MarkusH) [inline tests/serializers/test_natural.py]: Can you assert the values of the one object that exists in the db across imports, i.e. the pk didn't change. Also, I think it's warranted to have an assert before the first import that checks for an empty table.

File: tests/serializers/test_natural.py (hunk 3 above)
Comment (DmytryiStriletskyi) [inline tests/serializers/test_natural.py]: > Is there any reason why you test the same for both JSON and XML?

Yes. Bug, that i fixed, occurs for `xml` and `json` during `loaddata`.

**Comment by danielknell** [comment]
from a quick glance, when the user provides a PK in the fixture for a model with a UUID primary key, will this code not overwrite it?

File: tests/serializers/test_natural.py (hunk 4 above)
Comment (DmytryiStriletskyi) [inline tests/serializers/test_natural.py]: > Also, I think it's warranted to have an assert before the first import that checks for an empty table.

Why? Table by default is empty. If it isn't - `assert` (you suggest) won't passed. Better way is to clear table before first import. What do you think?

> Can you assert the values of the one object that exists in the db across imports, i.e. the pk didn't change.

Yes, I agree.

**Comment by DmytryiStriletskyi** [comment]
Do not understand your question clearly. Give me more case details.

Process is:
1. Fixture has no `pk`.
2. If fixture `natural key` exists, code get `pk` through `natural_key`.

If user provide `pk`, that exists in database, code will update this object in table.

**Comment by danielknell** [comment]
If the fixture provides a `pk`this code looks like it will overwrite it with the one from the database, it should only grab the value from the database if a pk is not set in the fixture.

**Comment by DmytryiStriletskyi** [comment]
@danielknell, it grab `pk` by `natural key` for updating object in next part of code by `pk`. I still do not understand what is wrong.

File: tests/serializers/test_natural.py (hunk 4 above)
Comment (MarkusH) [inline tests/serializers/test_natural.py]: What I mean by "assert the values of the one object that exists in the db across imports, i.e. the pk didn't change" is this:
```python
load_data()
old_obj = Foo.objects.get()  # This already asserts that there's exactly one object in the table
load_data()
new_object = Foo.objects.get()  # This again asserts a single objects in the table
self.assertEqual(old_obj.pk, new_obj.pk)
```

Comment (DmytryiStriletskyi) [inline tests/serializers/test_natural.py]: @MarkusH, i had a lot of time to understand better and did it. Anyway, thanks you! :)

**Comment by danielknell** [comment]
you always overwrite the value set here, if the user explicitly provided a pk then it should be respected.

https://github.com/django/django/blob/45bbb9aae0e7af6cd16fa4fe410219d4d65f829c/django/core/serializers/python.py#L98

**Comment by DmytryiStriletskyi** [comment]
> when the user provides a PK in the fixture for a model with a UUID primary key, will this code not overwrite it?

> you always overwrite the value set here, if the user explicitly provided a pk then it should be respected.

If user provides `pk`, block of code I present here (get `pk` by `natural_key`) won't execute (it won't do anything with `pk`). My answer is `yes.`: this code not overwrite it.

File: tests/serializers/test_natural.py (hunk 3 above)
Comment (DmytryiStriletskyi) [inline tests/serializers/test_natural.py]: > Sure, but the bug is not specific to either one. It's a general bug in how Django loads fixture data, regardless of the encoding format

Okay, i will leave only `json`.

**Comment by danielknell** [comment]
> If user provides pk, block of code I present here (get pk by natural_key) won't execute (it won't do anything with pk). My answer is yes.: this code not overwrite it.

its probably worth adding a test to prove that to be true.

**Comment by DmytryiStriletskyi** [comment]
@danielknell, sorry, but i missed. I kept in mind my [first implementation](https://github.com/DmytryiStriletskyi/django/blob/3100e73349d5e15deeb64f2d7364425dd7bc45f8/django/core/serializers/python.py#L103). Now it overwrite it, but i go ahead and fix it.

File: django/core/serializers/base.py [hunk 5]
```diff
@@ -221,7 +221,8 @@ def build_instance(Model, data, db):
     """
     default_manager = Model._meta.default_manager
 
-    if hasattr(default_manager, 'get_by_natural_key') and hasattr(Model, 'natural_key'):
+    if (data.get('id') is None and hasattr(default_manager, 'get_by_natural_key')
```
Comment (danielknell) [inline django/core/serializers/base.py]: might be worth using `Model._meta.pk.name` instead of explicit `'id'`

Comment (DmytryiStriletskyi) [inline django/core/serializers/base.py]: > might be worth using Model._meta.pk.name instead of explicit 'id'

I agree. Thanks you!

File: django/core/serializers/python.py [hunk 6]
```diff
@@ -100,6 +100,7 @@ def Deserializer(object_list, *, using=DEFAULT_DB_ALIAS, ignorenonexistent=False
                 data[Model._meta.pk.attname] = Model._meta.pk.to_python(d.get('pk'))
             except Exception as e:
                 raise base.DeserializationError.WithData(e, d['model'], d.get('pk'), None)
+
```
Comment (DmytryiStriletskyi) [inline django/core/serializers/python.py]: > Can you revert this change, please

Done

File: tests/serializers/test_natural.py (hunk 2 above)
Comment (DmytryiStriletskyi) [inline tests/serializers/test_natural.py]: > Please wrap docstrings at 79 chars.

Done.

**Comment by timgraham** [comment]
Can you make the test implementation match the other tests rather than using `call_command("loaddata", ...)`? Please rebase your branch and squash the commits. Uncheck "patch needs improvement" on the ticket when it's ready for review.

**Comment by timgraham** [comment]
Closing due to inactivity.
